### ISSUE TITLE
Explicitly list whta is exported in each file

### ISSUE THREAD

**Issue by kinow** [cylc/cylc#3087]
Bitten by a bug that could appear in linters, where one module `X` is importing an error `cylc.exceptions.CylcErrorAbc` to use itself, and another module `Y` is importing the error `from cylc.X import CylcErrorAbc`.

I suspect it would be possible to find this kind of issue through the [dunder `__all__`](https://www.python.org/dev/peps/pep-0008/#id44).

This would be useful too in case a user decides to do something like `from cylc.cycling.iso8601 import *`. Right now the user will get even classes from `isodatetime`, which is probably not expected/wanted by the user.

**Comment by hjoliver** [comment]
I think `__all__` only has a real effect in case of `from blah import *?  ...  which is normally discouraged, so we've never used it.  But I suppose, to suppose users using "cylc as a module" we should.  

This seems worth reading: http://xion.io/post/code/python-all-wild-imports.html

**Comment by hjoliver** [comment]
> Bitten by a bug in #3086 that could appear in linters, where ...

Is what you describe here actually a bug, or just bad form?

**Comment by hjoliver** [comment]
OK, just saw your comment here! https://github.com/cylc/cylc/pull/3086/files

**Comment by kinow** [comment]
>OK, just saw your comment here! https://github.com/cylc/cylc/pull/3086/files#r272818563

Yup! But you are right. Like that typed modules PEP-484, it has very little effect, as it is not enforced in runtime. In my IDE, if the module contains `__all__`, and I import anything that is not listed there, it adds a warning. But I believe `pycodestyle`, `black`, `flake8`, etc, should be able to catch that too (causing the issue to pop up in Codacy as well).

I think both `__all__` and the PEP-484 can be useful to help us providing an API that is easy to be used by other developers, as well as perhaps helping us maintain the code.

But like coverage, I think it is nice to have, but if we find a case in Cylc that it gets too complicated to add the `__all__`, or add types, then I think we can just leave it for later :+1:

**Comment by oliver-sanders** [comment]
I think the purpose of `__all__` is to clear out the cruft in user facing APIs and avoid nasty import issues when encountered.

Makes sense to use `__all__` at top level code (e.g. `cylc.flow`) or particular user interfaces (e.g. `cylc.flow.network.client`), I definitely wouldn't want to put `__all__` variables on *every* file though, that's just a maintenance headache.

**Comment by kinow** [comment]
>Makes sense to use __all__ at top level code (e.g. cylc.flow) or particular user interfaces (e.g. cylc.flow.network.client), I definitely wouldn't want to put __all__ variables on every file though, that's just a maintenance headache.

Fair enough :+1:
